jetstream/analysis.py

```diff
@@ -206,41 +217,125 @@ def _table_name(
+        window_period: AnalysisPeriod,
+        table_prefix=None,
+        analysis_basis=None,
+        metric_list: list[str] | None = None,
+    ):
         assert self.config.experiment.normandy_slug is not None
         normalized_slug = bq_normalize_name(self.config.experiment.normandy_slug)
         view_name = "_".join([normalized_slug, window_period.table_suffix])
-        wildcard_expr = "_".join([normalized_slug, window_period.value, "*"])
+        wildcard_expr = normalized_slug
 
         if analysis_basis:
             normalized_postfix = bq_normalize_name(analysis_basis)
             view_name = "_".join([normalized_slug, normalized_postfix, window_period.table_suffix])
-            wildcard_expr = "_".join(
-                [normalized_slug, normalized_postfix, window_period.value, "*"]
-            )
+            wildcard_expr = "_".join([normalized_slug, normalized_postfix])
 
         if table_prefix:
             normalized_prefix = bq_normalize_name(table_prefix)
             view_name = "_".join([normalized_prefix, view_name])
             wildcard_expr = "_".join([normalized_prefix, wildcard_expr])
 
+        prefix = f"{self.project}.{self.dataset}.{wildcard_expr}"
         sql = dedent(
             f"""
             CREATE OR REPLACE VIEW `{self.project}.{self.dataset}.{view_name}` AS (
                 SELECT
                     *,
-                    CAST(_TABLE_SUFFIX AS int64) AS window_index
-                FROM `{self.project}.{self.dataset}.{wildcard_expr}`
+                    CAST(REGEXP_EXTRACT(_TABLE_SUFFIX, r'[0-9]+$') AS INT64) AS window_index
+                FROM `{prefix}_{window_period.value}_*`
             )
             """
         )
+
+        if metric_list:
+            analysis_unit_list = [
+                "analysis_id",
+                "branch",
+                "enrollment_date",
+                "num_enrollment_events",
+                "analysis_window_start",
+                "analysis_window_end",
+                "exposure_date",
+                "num_exposure_events",
+            ]
+            analysis_unit_str = ",".join(analysis_unit_list)
+            using_list = analysis_unit_list
+            using_list.append("window_index")
+            using_str = ",".join(using_list)
+            join_sql = ""
+            join_clause = f" USING ({using_str}) "
+
+            # join all of the sub-queries together so that we get the client info
+            # combined with all of the metric columns in one view:
+            #   ```sql (snippet)
+            #   (
+            #       SELECT analysis_id, branch, # etc.
+            #           metric_name,
+            #           1 AS window_index
+            #       FROM `project.dataset.<experiment_slug>_<metric_name>_<period>_*`
+            #   ) FULL OUTER JOIN
+            #   (
+            #       SELECT analysis_id, branch, # etc.
+            #           other_metric_name,
+            #           1 AS window_index
```

Comment:
Is it possible that we end up with result tables for discrete metrics and result tables from the "normal" runs that both have results for the same window? E.g. if we run discrete metrics manually and then the scheduled jetstream runs are (re-)running for the same day? Could this create problems (if this is a possibility)? Just thinking out loud here...

Reply:
Yep, definitely a concern. I'd like to tackle this more comprehensively in a follow-up, but for now we should be ok because the tables should get deleted with a rerun.